modules/authglobus/lib/Auth/Source/Globus.php

```diff
@@ -0,0 +1,213 @@
+        curl_setopt($curl, CURLOPT_POSTFIELDS, OauthUtil::build_http_query($contents));
+
+        $newArr = curl_exec($curl);
+        $outInfo = curl_getinfo($curl);
+
+        return $newArr;
+    }
+
+    /**
+     * Retrieve user info/identity affiliations from the Globus API
+     *
+     * @param string $url The url + endpoints we want to hit
+     * @param array $token The bearer who will make the request
+     */
+    private function getGlobusInfo($url, $token) {
+        $header = 'Authorization: Bearer ' . $token;
+        $options = array(
+            'http' => array(
+                'method' => 'GET',
+                'header' => $header
+            )
+        );
+
+        $context = stream_context_create($options);
+        $info = file_get_contents($url, false, $context);
+
+        return json_decode($info, true);
+    }
+
+    /**
+     * Obtain an Authorization Code
+     *
+     * @param array &$state  Information about the current authentication.
+     */
+    public function authenticate(&$state)
+    {
+        // We are going to need the authId in order to retrieve this authentication source later
+        $state[self::AUTHID] = $this->authId;
+
+        $stateID = SimpleSAML_Auth_State::getStateId($state);
+        SimpleSAML\Logger::debug("globus auth at state = " . $stateID);
+
+        $consumer = new sspmod_oauth_Consumer($this->key, $this->secret);
+
+        $request = array(
+            'client_id' => $this->key,
+            'scope' => $this->scope,
+            'state' => $stateID,
+            'response_type' => 'code',
+            'redirect_uri' => $this->redirect_uri
+        );
+
+        $state['globusauth:request'] = $request;
+        $urlAppend = \SimpleSAML\Utils\HTTP::addURLParameters(API_ENDPT . '/oauth2/authorize', $request);
+        SimpleSAML_Auth_State::saveState($state, self::STAGE_INIT);
+
+        // Get an Authorization Code from Globus
+        /* Example: https://auth.globus.org/v2/oauth2/authorize?client_id=69ba5e62-7285-45db-952d-e0bb73b5eac7&scope=urn:globus:auth:scope:transfer.api.globus.org:all urn:globus:auth:scope:auth.globus.org:view_identities offline_access&response_type=code&redirect_uri=https://www.example.org/my_app/login&state=g6l14b2xlgx4dtce8d2ja714i
+          */
+        $authorizeUrl = $consumer->getAuthorizeRequest($urlAppend, $request);
+    }
+
+    /**
+     * Exchange authorization code for an access token
+     *
+     * @param array &$state  Information about the current authentication.
+     */
+    public function finalStep(&$state)
+    {
+        $request = array(
+            'code' => $_GET['code'],
+            'redirect_uri' => $this->redirect_uri,
+            'grant_type' => 'authorization_code'
+        );
+        $oauthSignature = $this->key . ":" . $this->secret;
+
+        // Exchange the code we got earlier for an access token
+        $result = json_decode($this->getGlobusTokens($request, $oauthSignature));
```

Comment:
what happens if this fails?  looks like we are assuming that it is going to be able to get an auth token, we should account for the globus api not being reachable or failing to respond how we expect it to and possibly use ` SimpleSAML_Auth_State::throwException` so that we get better looking error messages than white background black text with php errors